dev/tests/api-functional/testsuite/Magento/GraphQl/Quote/Customer/SetShippingMethodsOnCartTest.php

```diff
@@ -49,46 +63,264 @@ protected function setUp()
         $this->quoteFactory = $objectManager->get(QuoteFactory::class);
         $this->quoteIdToMaskedId = $objectManager->get(QuoteIdToMaskedQuoteIdInterface::class);
         $this->customerTokenService = $objectManager->get(CustomerTokenServiceInterface::class);
+        $this->rate = $objectManager->get(Rate::class);
+        $this->productRepository = $objectManager->get(ProductRepositoryInterface::class);
     }
 
+    /**
+     * @magentoApiDataFixture Magento/Checkout/_files/quote_with_virtual_product_and_address.php
+     * @magentoApiDataFixture Magento/Checkout/_files/enable_all_shipping_methods.php
+     * @throws \Exception
+     */
     public function testShippingMethodWithVirtualProduct()
     {
-        $this->markTestIncomplete('https://github.com/magento/graphql-ce/issues/423');
+        $maskedQuoteId = $this->getMaskedQuoteIdByReversedQuoteId('test_order_with_virtual_product');
+
+        /** @var Quote $quote */
+        $quote = $this->quoteFactory->create();
+        $this->quoteResource->load($quote, 'test_order_with_virtual_product', 'reserved_order_id');
+
+        $shippingAddress = $quote->getShippingAddress();
+        $rate = $this->rate;
+
+        $rate->setPrice(2)
+            ->setAddressId($shippingAddress->getId())
+            ->setCode('flatrate_flatrate');
+        $shippingAddress->setShippingMethod('flatrate_flatrate')
+            ->addShippingRate($rate)
+            ->save();
```

Comment:
Method `save` is deprecated. To avoid usage of deprecated methods just try to use RepositoryInterface instead of Models, ResourceModels and Factories in tests. As far as I know it is not a rule it's good practice. No need to change until not requested by reviewers.